text/0000-feature-metadata.md

```diff
@@ -0,0 +1,310 @@
+information is in a separate RFC.
+
+# Motivation
+
+[motivation]: #motivation
+
+Features are widely used as a way to do things like reduce dependency count,
+gate `std` or `alloc`-dependent parts of code, or hide unstable API. Use is so
+common that many larger crates wind up with tens of feature gates, such as
+[`tokio`] with 24. Despite being a first class component of crate structure,
+there are some limitations that don't have elegant solutions:
+
+-   Documentation is difficult, often requiring library authors to manually manage
+    a table of descriptions
+-   There is no way to deprecate old features, as a way to help crates maintain
+    semvar compliance
+-   Features cannot be hidden from use in any way
+
+This RFC proposes a plan that add that information to `Cargo.toml`, solving
+these problems.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Usage is simple: features will be able to be specified as a table, instead of
+just a dependency array. This sample section of `Cargo.toml` shows new
+possibilities:
+
+```toml
+[features]
+# Current configuration will continue to work
+foo = []
+# New configurations
+bar = { requires = ["foo"], doc = "simple docstring here"}
+baz = { requires = ["foo"], public = false}
+qux = { requires = [], deprecated = true }
+quux = { requires = [], deprecated = { since = "1.2.3", note = "don't use this!" } }
+
+# Features can also be full tables if descriptions are longer
+[features.corge]
+requires = ["bar", "baz"]
+doc = """
+# corge
+
+This could be a longer description of this feature
+"""
+```
+
+The following keys would be allowed in a feature table:
+
+-   `requires`: This is synonymous with the existing array describing required
+    features. For example, `foo = ["dep:serde", "otherfeat"]` will be identical to
+    `foo = { requires = ["dep:serde", "otherfeat"] }`
+-   `doc`: A markdown docstring describing the feature. Like with `#[doc(...)]`,
+    the first line will be treated as a summary.
+-   `deprecated`: This can be either a simple boolean, a string, or an object
+    with `since` and/or `note` keys. Cargo will warn downstream crates using this
+-   `public`: A boolean flag defaulting to `true` that indicates whether or not
+    downstream crates should be allowed to use this feature
+
+If a downstream crate attempts to use the features `baz` and `qux`, they will
+see messages like the following:
+
+```
+warning: feature `quux` on crate `mycrate` is deprecated since version
+  1.2.3: "don't use this!"
+
+error: feature `baz` on crate `mycrate` is private and cannot be used by
+  downstream crates
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Validation and parsing of the new schema, described above, shoud be relatively
+straightforward. Each of the added keys is discussed individually in the
+following sections:
+
+## `requires`
```

Comment:
That is sort-of an inverse relationship though. Maybe this should be `activates` or something similar since it doesn't _require_ the other features to be active but instead _activates_ them when this is active.

Reply:
My 2¢ would be that the key names should describe/modify the feature they're a member of (such that you could read "X has [key] [value]" or "X is [key]" or "X [key] [values]"). So if you have:

```toml
X = { requires = ["Y"] }
```

That works because `requires` describes X, i.e. "Feature X `requires` feature Y". I think `required` would sound a bit more like feature X is in some way required by something, like a T/F flag ("X is required"). `required-features` works because `required` modifies the other `features` and not X itself (so those depended upon by X ("X has `required-features` Y"). So, I think either `requires` or `required-features` would work best here, `required-features` just seemed a bit long (though I'm not opposed to it).

@Nemo157's suggestion could be reasonable too (feature X `activates` feature Y), or something else like `enables`. I'll defer to you for what terminology would be most consistent.

Reply:
Another potential term is "enable".  https://github.com/rust-lang/rfcs/pull/3374 uses that (note: it isn't approved either)

Reply:
I think I like that one the best, and it's definitely the most common language - it is the verb used throughout https://doc.rust-lang.org/cargo/reference/features.html as well.

I updated `requires` -> `enables` and will resolve this